build_tools/python/e2e_test_framework/definitions/common_definitions.py

```diff
@@ -122,6 +131,29 @@ class DeviceSpec(object):
   # these parameters to set up the devices. E.g. set CPU mask.
   device_parameters: List[str] = dataclasses.field(default_factory=list)
 
+  def __str__(self):
+    return self.name
+
+  @staticmethod
+  def build(id: str,
+            device_name: str,
+            tags: Sequence[str],
+            host_environment: HostEnvironment,
+            architecture: DeviceArchitecture,
+            device_parameters: Optional[Sequence[str]] = None):
+    tag_part = tags = ",".join(tags)
+    # Format: <device_name>[<tag>,...]
+    name = f"{device_name}[{tag_part}]"
+    device_parameters = [] if device_parameters is None else list(
+        device_parameters)
```

Comment:
Maybe this is simpler?

```suggestion
    device_parameters = device_parameters or []
```

Reply:
Ah, nice.